### ISSUE TITLE
Evaluation running slowly

### ISSUE THREAD

**Issue by bnic24** [deepset-ai/FARM#68]
I've noticed that eval runs more slowly than training, which doesn't seem intuitive to me. Shouldn't eval run faster as there's no back propagation?

For example, while training a standard LM:

Training: `25941/25941 [3:31:49<00:00,  2.95it/s]` = 2.4 it/s
Eval:  `3811/3811 [41:24<00:00,  1.53it/s]` = 1.53 it/s

**Comment by Timoeller** [comment]
Evaluation is doing a lot of computations on the CPU. I do not see straightforward ways of doing all kinds of evaluations on GPUs.

If evaluation takes too long we decrease how often it is called by a parameter value "eval_every". Maybe adjust this to your dataset size, since eval is called every "eval_every" steps.

**Possible solution**: FARM Eval is not yet optimized for speed. Especially the aggregation of all predictions into huge lists (of numpy arrays) seems wasteful. If you have any thoughts on how to improve this for all downstream tasks, we are open to suggestions.

**Comment by bnic24** [comment]
I've adjusted the "evaluate_every" param but I'd still like to eval at least after every epoch and even only doing that takes up about 20% of the total training time!

Ah, that would explain it. I naively assumed that evaluation was already running on the GPUs.

Instead of having massive preds_all and label_all arrays, would it be possible just to store the preds and labels for the current batch and then calculate and store the metrics for the batch (and then take the average across batches to produce the final metrics)? However might might just reduce memory pressure, and not improve the speed. And I'm not sure how it would work for reports.

I think one motivation for having the massive arrays was to help with debugging. If that's the case then maybe we could introduce a special debug mode (or a quick_eval mode).

**Comment by Timoeller** [comment]
I think that is the main problem: "And I'm not sure how it would work for reports."

The evaluation is currently written to work on CPU, be it the report or compute_metrics(). We would need to rewrite both to work in pytorch, call these functions on batches and aggregate over batches.

Fast evaluation on lots of data seems especially relevant for LM finetuning, so you could start by implementing this "GPU eval" only for LM finetuning?

@brandenchan can you please have a look how to best implement this inside the Evaluator? I guess we do not have metrics for LM finetuning, just the LM loss and next sentence loss that needs aggregation.

**Comment by Timoeller** [comment]
Hey @bnic24 We have also conducted some experiments with LM adaptation now and realized we do not need so large dev sets for evaluating the models performance. So making the evaluator more efficient isn't high priority any more. 
I will close this issue and update you if we should updat the code for evaluating.
